File: EcoSwapAndroid/src/main/java/com/darrenthiores/ecoswap/android/presentation/login/LoginScreen.kt
package com.darrenthiores.ecoswap.android.presentation.login

import androidx.compose.foundation.Image
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.darrenthiores.ecoswap.android.R
import com.darrenthiores.ecoswap.android.presentation.components.buttons.DefaultButton
import com.darrenthiores.ecoswap.android.presentation.components.textfields.DefaultTextField
import com.darrenthiores.ecoswap.android.presentation.components.textfields.PasswordTextField
import com.darrenthiores.ecoswap.android.theme.Blue
import com.darrenthiores.ecoswap.android.theme.EcoSwapTheme
import com.darrenthiores.ecoswap.android.theme.FootnoteR
import com.darrenthiores.ecoswap.android.theme.SubHeadlineB
import com.darrenthiores.ecoswap.android.theme.SubHeadlineR
import com.darrenthiores.ecoswap.android.theme.Title2B
import com.darrenthiores.ecoswap.presentation.login.LoginEvent
import com.darrenthiores.ecoswap.presentation.login.LoginState

@Composable
fun LoginScreen(
    state: LoginState,
    onEvent: (LoginEvent) -> Unit,
    onSignUp: () -> Unit
) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(
                horizontal = 24.dp,
                vertical = 56.dp
            ),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Image(
            modifier = Modifier
                .size(128.dp),
            painter = painterResource(id = R.drawable.ecoswap_logo),
            contentDescription = stringResource(id = R.string.ecoswap_logo),
            contentScale = ContentScale.FillHeight
        )

        Spacer(modifier = Modifier.height(24.dp))

        Text(
            text = stringResource(id = R.string.login_title),
            style = Title2B
        )

        Text(
            text = stringResource(id = R.string.login_headline),
            style = SubHeadlineR.copy(
                color = Color.Gray
            )
        )

        Spacer(modifier = Modifier.height(24.dp))

        DefaultTextField(
            text = state.email,
            onTextChange = {
                onEvent(LoginEvent.OnEmailChange(it))
            },
            placeholder = stringResource(id = R.string.email_placeholder)
        )

        Spacer(modifier = Modifier.height(16.dp))

        PasswordTextField(
            text = state.password,
            onTextChange = {
                onEvent(LoginEvent.OnPasswordChange(it))
            },
            placeholder = stringResource(id = R.string.password_placeholder),
            isVisible = state.showPassword,
            onToggle = {
                onEvent(LoginEvent.ToggleShowPassword)
            }
        )

        Spacer(modifier = Modifier.height(16.dp))

        Row(
            modifier = Modifier
                .fillMaxWidth()
        ) {
            Spacer(modifier = Modifier.weight(1f))

            Text(
                modifier = Modifier
                    .clickable {

                    },
                text = stringResource(id = R.string.forget_password),
                style = FootnoteR.copy(
                    color = Blue
                )
            )
        }

        Spacer(modifier = Modifier.height(32.dp))

        DefaultButton(
            modifier = Modifier
                .fillMaxWidth(),
            label = stringResource(id = R.string.sign_in),
            onClick = {
                onEvent(LoginEvent.Login)
            },
            disabled = state.isLoading || state.email.isEmpty() ||
            state.password.isEmpty()
        )

        Spacer(modifier = Modifier.height(16.dp))

        Row(
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = stringResource(id = R.string.dont_have_acc),
                style = SubHeadlineR
            )

            Text(
                modifier = Modifier
                    .clickable {
                        onSignUp()
                    },
                text = stringResource(id = R.string.sign_up),
                style = SubHeadlineB.copy(
                    color = Blue
                )
            )
        }
    }
}

@Preview
@Composable
private fun LoginScreenPreview() {
    EcoSwapTheme {
        LoginScreen(
            state = LoginState(),
            onEvent = {  },
            onSignUp = {  }
        )
    }
}

File: EcoSwapiOS/EcoSwapiOS/Presentation/Login/LoginScreen.swift
//
//  LoginScreen.swift
//  EcoSwapiOS
//
//  Created by Darren Thiores on 02/01/24.
//  Copyright © 2024 orgName. All rights reserved.
//

import SwiftUI

struct LoginScreen: View {
    @StateObject private var viewModel = IosLoginViewModel()
    
    var body: some View {
        NavigationStack {
            VStack(alignment: .center, spacing: 0) {
                Image("ecoswap_logo")
                    .resizable()
                    .frame(
                        width: 128,
                        height: 128
                    )
                    .scaledToFit()
                
                Group {
                    Spacer()
                        .frame(height: 24)
                    
                    Text("Welcome EcoSwap 👋🏻")
                        .font(.Title2B)
                    
                    Text("Enter your Email & Password to Sign in")
                        .font(.SubHeadlineR)
                        .foregroundColor(.gray)
                    
                    Spacer()
                        .frame(height: 24)
                }
                
                Group {
                    DefaultTextField(
                        text: Binding(
                            get: {
                                viewModel.state.email
                            },
                            set: {
                                viewModel.onEvent(
                                    event: .OnEmailChange(email: $0)
                                )
                            }
                        ),
                        placeholder: "Your email"
                    )
                    
                    Spacer()
                        .frame(height: 16)
                    
                    PasswordTextField(
                        text: Binding(
                            get: {
                                viewModel.state.password
                            },
                            set: {
                                viewModel.onEvent(
                                    event: .OnPasswordChange(password: $0)
                                )
                            }
                        ),
                        passwordVisible: Binding(
                            get: {
                                viewModel.state.showPassword
                            },
                            set: { _ in
                                viewModel.onEvent(
                                    event: .ToggleShowPassword()
                                )
                            }
                        ),
                        placeholder: "Password"
                    )
                    
                    Spacer()
                        .frame(height: 16)
                }
                
                HStack {
                    Spacer()
                    
                    Button {
                        
                    } label: {
                        Text("Forget Password?")
                            .font(.FootnoteR)
                            .foregroundColor(.Blue)
                    }
                    .buttonStyle(.plain)
                }
                
                Spacer()
                    .frame(height: 32)
                
                Group {
                    ButtonFill(
                        label: "Sign In",
                        onClick: {
                            viewModel.onEvent(
                                event: .Login()
                            )
                        },
                        disabled: viewModel.state.isLoading || viewModel.state.email.isEmpty || viewModel.state.password.isEmpty,
                        fillWidth: true,
                        isLoading: viewModel.state.isLoading
                    )
                    
                    Spacer()
                        .frame(height: 16)
                    
                    HStack(spacing: 0) {
                        Spacer()
                        
                        Text("Don’t Have an Account? ")
                            .font(.SubHeadlineR)
                        
                        NavigationLink {
                            RegisterScreen()
                        } label: {
                            Text("Sign Up")
                                .font(.SubHeadlineB)
                                .foregroundColor(.Blue)
                        }
                        .buttonStyle(.plain)
                        
                        Spacer()
                    }
                }
                
                Spacer()
            }
            .padding(.horizontal, 24)
            .padding(.vertical, 56)
            .onAppear {
                viewModel.startObserving()
            }
            .onDisappear {
                viewModel.dispose()
            }
        }
    }
}

#Preview {
    LoginScreen()
}
